### ISSUE TITLE
doctor --config: widen the MLX table to the remaining nine CUDA-only fields

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#958]
> **CORRECTION, 2026-09-15 — the ask below was wrong and this paragraph supersedes it.**
> I filed this asking for **all nine** fields and `None of the 18`. Five of the nine
> **cannot be loaded on `backend: mlx` at all** — the config is refused before
> `doctor --config` ever sees it — so a table row for them is a row nobody can reach.
> Verified by loading each of the nine through the real loader:
>
> | field | loads on `backend: mlx`? | belongs in the table |
> |---|---|---|
> | `training.use_mod` | parses | **yes** |
> | `training.moe_lora` | parses | **yes** |
> | `training.quantization_aware` | parses | **yes** |
> | `training.use_fsdp2_compile` | parses | **yes** |
> | `training.use_cut_ce` | REFUSED — v0.28.0 features not supported on mlx | no |
> | `training.kernel_auto_compose` | REFUSED — same gate | no |
> | `training.use_longlora` | REFUSED — LongLoRA is not supported on mlx | no |
> | `training.activation_offloading` | REFUSED — v0.28.0 gate | no |
> | `training.packing_cross_doc_attn_mask` | REFUSED — never mapped to a valid TRL packing strategy | no |
>
> **So the ask is four fields, and the count goes 9 -> 13, not 9 -> 18.** The
> acceptance boxes at the bottom are rewritten accordingly; the original text is
> kept below the line so the history reads honestly. #961 implements the corrected
> scope. Anyone arriving here from #965 or elsewhere should work from the corrected
> boxes, not the original ones.

---

`soup doctor --config` now scopes its all-clear to the fields it actually knows about (#911, fixing #903):

```
None of the 9 setting(s) known to be unread on task=sft backend=mlx is set in this config.
```

That sentence is honest, but the number is only as strong as the table behind it, and the table is currently 9 entries wide on the one reviewed pair. #903 named eleven CUDA-only fields; #911 added two of them, deliberately and in the open. The others are listed in the correction table above.

All of them are real `TrainingConfig` fields with **zero hits across the MLX modules** (`trainer/mlx_sft.py`, `trainer/mlx_optim.py`, `trainer/mlx_masking.py`), so any of them would pass `test_issue755_backend_support_registry.py`'s bidirectional guard — but passing that guard is not sufficient, because a field the schema refuses on this backend can never reach the report.

### Why this is worth doing

`None of the 9` reads stronger than 9-of-20 warrants. Widening the table is the only thing that makes the count mean what a reader takes it to mean, and every reachable entry is cheap: one `SupportEntry`, one `_check_unsupported` branch, one grep to prove zero hits.

### Two notes carried over from the #911 review

1. **`use_liger` is not purely "ignored".** `commands/train.py:1193` calls `validate_liger_config`, which errors whenever `device != "cuda"` and exits 1 — so on a real Apple Silicon box `soup train` refuses the run before MLX ever sees the field. `IGNORED` is still defensible under this module's definitions (MLX genuinely never reads it, and `doctor --config` does not run that validation), but a clause in the reason string would make the report match what the user will actually experience. `neftune_alpha` is the pure case.

   **Correction, 2026-09-15:** I later extended this note to `quantization_aware` as well. That was wrong — `validate_qat_config` has **no mlx branch at all** (it refuses `unsloth` only), so the clause would be false there. It applies to `use_fsdp2_compile`, where `validate_fsdp2_compile_config(..., backend="mlx")` refuses unconditionally and `commands/train.py:1186` exits before `resolve_trainer`.

2. **A fixture in `tests/test_issue903_*` depends on its own literal indentation.** The `"  use_liger: true\n              neftune_alpha: 5"` string parses correctly today — I reproduced it, `model_fields_set == ['neftune_alpha', 'use_liger']` — but only because the 14 spaces survive `dedent`. Any future edit there breaks it into invalid YAML silently rather than failing loudly. `"  use_liger: true\n  neftune_alpha: 5"` is equivalent and cannot break that way.

### Acceptance (corrected 2026-09-15)

- [ ] The **four reachable** fields present in `_MLX_SFT` with a reason string each, every one justified by a zero-hit grep stated in the PR: `use_mod`, `moe_lora`, `quantization_aware`, `use_fsdp2_compile`
- [ ] The five unreachable fields explicitly **left out**, and the PR says why — a row for a config the schema refuses is a row nobody can reach
- [ ] Matching `trainer/mlx_sft.py::_check_unsupported` branches so the trainer warns too, not only the doctor
- [ ] `soup doctor --config` on a clean mlx config reports `None of the 13 setting(s) ...`
- [ ] A test that sets several of the new fields at once and asserts the table rows and the count
- [ ] The two notes above addressed or explicitly declined with a reason, taking note of the correction to note 1

<details>
<summary>Original acceptance criteria, superseded — kept for the record</summary>

- [ ] All nine fields present in `_MLX_SFT` with a reason string each, every one justified by a zero-hit grep stated in the PR
- [ ] Matching `trainer/mlx_sft.py::_check_unsupported` branches so the trainer warns too, not only the doctor
- [ ] `soup doctor --config` on a clean mlx config reports `None of the 18 setting(s) ...`
- [ ] A test that sets several of the new fields at once and asserts the table rows and the count
- [ ] The two notes above addressed or explicitly declined with a reason

</details>

**Comment by MakazhanAlpamys** [comment]
**Correcting this issue. I wrote it, and it is wrong in two places.** @ege-arhan followed it faithfully in #961 and was walked into both, so the correction belongs here rather than in the PR thread.

## 1. The premise is insufficient — five of the nine cannot be reported at all

I argued that all nine fields have zero hits in the three MLX modules and would therefore pass the bidirectional guard. Both halves are true, and both are beside the point: **I never checked whether a config setting them can be loaded.** Five cannot.

```
REFUSED  use_cut_ce                   v0.28.0 features are not supported on the Apple Silicon mlx backend
REFUSED  kernel_auto_compose          (same gate)
REFUSED  use_longlora                 LongLoRA is not supported on the mlx backend
REFUSED  activation_offloading        (same v0.28.0 gate)
REFUSED  packing_cross_doc_attn_mask  never mapped to a valid TRL packing_strategy - refused on EVERY backend
PARSES   use_mod, moe_lora, quantization_aware, use_fsdp2_compile
```

`schema.py:4341-4368` (`_validate_v028_speed_memory_supported_tasks`) and `schema.py:3509-3518` (`_validate_cross_doc_attn_mask`).

So a config setting any of those five is refused before `soup doctor --config` runs, and a `_check_unsupported` branch for them is dead code.

**The acceptance box asking for `None of the 18 setting(s) ...` is therefore the wrong ask** — 18 would be the dishonest number, which is exactly what #903 and this issue exist to prevent. The reachable count is 4:

```
soup doctor --config mlx4.yaml     # use_mod, moe_lora, quantization_aware, use_fsdp2_compile
4 setting(s) written here are not read on backend=mlx.
```

**Revised acceptance:**

- [ ] The four reachable fields — `use_mod`, `moe_lora`, `quantization_aware`, `use_fsdp2_compile` — are in `_MLX_SFT` with matching `_check_unsupported` branches
- [ ] The five unreachable ones are either omitted, or carry a reason string saying the config is **refused before the doctor runs**, which is different information from "MLX ignores this" and the user hitting it deserves the right one
- [ ] A clean mlx config reports the reachable count, whatever that number turns out to be
- [ ] A test setting several of them at once asserts the rows and the count
- [ ] `use_fsdp2_compile` and `quantization_aware` get the same "`soup train` refuses on non-CUDA before MLX sees it" clause as `use_liger` — `validate_fsdp2_compile_config` at `commands/train.py:1172-1184` and `validate_qat_config` at `:1143-1153` both exit 1 before `resolve_trainer` at `:1398`

## 2. Note (b) was actively wrong, and following it breaks a test

I suggested replacing `"\n              neftune_alpha: 5"` with a 2-space version, calling the 14 spaces fragile. They are load-bearing. The `config_at` fixture (`tests/test_issue755_backend_support_registry.py:53-66`) interpolates `textwrap.indent(training, "  ")` at a 12-space-indented position **inside a `textwrap.dedent` f-string**, so only line 1 receives the template's indentation and continuation lines must carry it literally. A 2-space line 2 lands at column 4, becomes `dedent`'s common prefix, and shreds the document into a YAML parser error.

The fix belongs in the fixture — `textwrap.indent(textwrap.dedent(training), "  ")` — not at the call site. Striking note (b); I will fix the fixture separately.

## 3. A real find that came out of #961, now #964

`use_fsdp2_compile` is the **first registry field with a digit in its name**, and the guard's extractor at `tests/test_issue755_backend_support_registry.py:150` is `re.findall(r"\b(?:training|data)\.[a-z_]+", ...)`. `[a-z_]` excludes digits, so the field is harvested as `training.use_fsdp` and the guard fails on a correct entry. That blocks this issue and is tracked separately.

## What this issue is really about, restated

The guard checks whether a field is **mentioned** in the MLX modules. It does not check whether a config carrying that field can **reach** them. Those are different questions, the second is the one a user experiences, and I conflated them when filing this. Anyone picking this up should treat reachability as the first check, not the last.

#961 by @ege-arhan is open against this and is assigned. Two of its four blocking items are mine.

**Comment by MakazhanAlpamys** [comment]
**Claimed by @ege-arhan**, working in #961.

Recorded here rather than only in the pull request, because the assignee field silently refused: GitHub accepts an assignee only if they have push access or have commented on **this issue**, and `gh issue edit --add-assignee` exits 0 either way. That is how 11 claims ended up invisible today, including this one.

So the `claimed` label plus this comment is the registry. If you are looking for unclaimed work, trust the label and the comments, not an empty assignee field.

**Comment by MakazhanAlpamys** [comment]
Corrected the acceptance criteria above, and flagged it here so anyone subscribed sees the change rather than finding a silently-edited issue.

**The ask was wrong: it is four fields, not nine, and the count goes 9 -> 13, not 9 -> 18.** Five of the nine I listed cannot be loaded on `backend: mlx` at all — the config is refused before `doctor --config` ever sees it — so a table row for them is a row nobody can reach. I verified this by loading each of the nine through the real loader rather than by reading the gates:

```
use_mod, moe_lora, quantization_aware, use_fsdp2_compile   -> parse
use_cut_ce, kernel_auto_compose, activation_offloading     -> REFUSED (v0.28.0 gate)
use_longlora                                               -> REFUSED (LongLoRA not supported on mlx)
packing_cross_doc_attn_mask                                -> REFUSED
```

I also got a second thing wrong and corrected it in note 1: I extended the pre-trainer-refusal clause to `quantization_aware`, but `validate_qat_config` has **no mlx branch at all** — it refuses `unsloth` only. What I saw locally was a missing-torchao artifact of my box, not a backend fact. The clause belongs on `use_fsdp2_compile`, where the refusal really is unconditional.

The original criteria are kept in a collapsed block rather than deleted, so the history reads honestly.

@ege-arhan's #961 implements the corrected scope and is the reason this came to light — they built the four-field version and explained why the other five were dropped. This issue text is what would have sent the next person down the wrong path, including @dchaudhari7177 on the adjacent #965.

**Comment by MakazhanAlpamys** [comment]
**Held by @ege-arhan — recording it here because the assignee field will not take it.**

`gh issue edit --add-assignee ege-arhan` returns success and silently assigns nobody: GitHub only accepts an assignee who is a repository collaborator or who has commented on the issue itself, and @ege-arhan's work on this is all on PR #961. So the API reports `assignees: []` no matter how many times I run it. The `claimed` label is the part that actually holds, and this comment is the name behind it.

That matters beyond bookkeeping: on 2026-09-14 I told @dchaudhari7177 that #965 was held because "#958 is assigned to @ege-arhan", and the API did not agree with me. @ege-arhan confirmed on 09-15 04:39 that they want to keep it — *"I will continue — please keep it with me"* — well inside the deadline I set, so **the hold stands and #965 is being told separately.**

PR #961 implements the corrected four-field scope and is green on its latest head.
